scripts/src/java/org/oppia/android/scripts/todo/TodoCheck.kt

```diff
@@ -0,0 +1,302 @@
+ *
+ * @param todos the list of all the failure causing todos
+ * @param todoExemptionList the list contating the todo exemptions
+ * @param repoPath path of the repo to be analyzed
+ * @return list obtained after filtering the exemptions
+ */
+private fun retrieveTodosAfterExemption(
+  todos: List<Todo>,
+  todoExemptionList: List<TodoExemption>,
+  repoPath: String
+): List<Todo> {
+  return todos.filter { todo ->
+    !todoExemptionList.any { it ->
+      it.exemptedFilePath == todo.filePath.removePrefix(repoPath) &&
+        todo.lineNumber in it.getLineNumberList()
+    }
+  }
+}
+
+/**
+ * Retrieves a list of redundant exemptions.
+ *
+ * @param todos the list of all the failure causing todos
+ * @param todoExemptionList the list contating the todo exemptions
+ * @param repoPath path of the repo to be analyzed
+ * @return a list of all the redundant exemptions
+ */
+private fun retrieveRedundantExemptions(
+  todos: List<Todo>,
+  todoExemptionList: List<TodoExemption>,
+  repoPath: String
+): List<Pair<String, Int>> {
+  return todoExemptionList.flatMap { exemption ->
+    exemption.getLineNumberList().mapNotNull { exemptedLineNumber ->
+      if (
+        !todos.any {
+          it.filePath.removePrefix(repoPath) == exemption.exemptedFilePath &&
+            it.lineNumber == exemptedLineNumber
+        }
+      ) {
+        Pair(exemption.exemptedFilePath, exemptedLineNumber)
+      } else {
+        null
+      }
+    }
+  }
+}
+
+/**
+ * Checks whether a line of code contains a poorly formatted todo.
+ *
+ * @param codeLine the line of code to be checked
+ * @param todoRegex regex of todo
+ * @param todoStartingRegex regex for the starting of the todo
+ * @param correctTodoFormatRegex regex of the correct todo format
+ * @return whether the line contains a poorly formatted TODO
+ */
+private fun checkIfTodoIsPoorlyFormatted(
+  codeLine: String,
+  todoRegex: Regex,
+  todoStartingRegex: Regex,
+  correctTodoFormatRegex: Regex
+): Boolean {
+  if (todoStartingRegex.containsMatchIn(codeLine)) {
+    if (!correctTodoFormatRegex.containsMatchIn(codeLine)) {
+      return true
+    }
+    return false
+  }
+  if (todoRegex.containsMatchIn(codeLine)) {
+    return true
+  }
+  return false
+}
+
+/**
+ * Checks whether a todo does not corresponds to an open issue on GitHub.
+ *
+ * @param codeLine the line of code to be checked
+ * @param openIssueList the list of all the open issues of this repository on GitHub
+ * @param correctTodoFormatRegex regex of the correct todo format
+ * @return whether the todo does not corresponds to an open issue
+ */
+private fun checkIfOpenIssueFailure(
+  codeLine: String,
+  openIssueList: List<Issue>,
+  correctTodoFormatRegex: Regex
+): Boolean {
+  if (!correctTodoFormatRegex.containsMatchIn(codeLine)) {
+    return false
+  }
+  val match = correctTodoFormatRegex.find(codeLine)
+  val parsedIssueNumberFromTodo = match?.groupValues?.get(1)
```

Comment:
I'm wondering a bit if it'd be simpler to move all the TODO parsing bits over to TodoCollector so that all the regex parsing, data class representations, and logic exists in one place.

Reply:
Revamption done.